checker/src/org/checkerframework/checker/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,770 @@
+                return true;
+            } else if (AnnotationUtils.areSameByClass(rhs, LTEqLengthOf.class)) {
+                return false;
+            }
+
+            // Neither is LTEL. Both must be LTOM, LTL, or Bottom. And the two must be
+            // different. The only way this can return true is if rhs is Bottom, or if rhs is
+            // LTOM and lhs is LTL.
+            if (AnnotationUtils.areSameByClass(rhs, UpperBoundBottom.class)) {
+                return true;
+            }
+
+            if (AnnotationUtils.areSameByClass(rhs, LTOMLengthOf.class)
+                    && AnnotationUtils.areSameByClass(lhs, LTLengthOf.class)) {
+                return true;
+            }
+
+            // Every other case results in false.
+            return false;
+        }
+    }
+
+    @Override
+    public TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                new UpperBoundTreeAnnotator(this), super.createTreeAnnotator());
+    }
+
+    protected class UpperBoundTreeAnnotator extends TreeAnnotator {
+
+        public UpperBoundTreeAnnotator(UpperBoundAnnotatedTypeFactory factory) {
+            super(factory);
+        }
+
+        /**
+         * This exists specifically for Math.min. We need to special case Math.min because it has
+         * unusual semantics: it can be used to combine annotations for the UBC, so it needs to be
+         * special cased. Do not special case other methods here unless you have a compelling reason
+         * to do so.
+         */
+        @Override
+        public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror type) {
+            ExecutableElement fcnMin =
+                    TreeUtils.getMethod("java.lang.Math", "min", 2, processingEnv);
+            if (TreeUtils.isMethodInvocation(tree, fcnMin, processingEnv)) {
+                AnnotatedTypeMirror leftType = getAnnotatedType(tree.getArguments().get(0));
+                AnnotatedTypeMirror rightType = getAnnotatedType(tree.getArguments().get(1));
+
+                type.replaceAnnotation(
+                        qualHierarchy.greatestLowerBound(
+                                leftType.getAnnotationInHierarchy(UNKNOWN),
+                                rightType.getAnnotationInHierarchy(UNKNOWN)));
+            }
+            return super.visitMethodInvocation(tree, type);
+        }
+
+        @Override
+        public Void visitMemberSelect(MemberSelectTree tree, AnnotatedTypeMirror type) {
+            if (tree.getIdentifier().contentEquals("length")
+                    && InternalUtils.typeOf(tree.getExpression()).getKind() == TypeKind.ARRAY) {
+                String arrName =
+                        FlowExpressions.internalReprOf(this.atypeFactory, tree.getExpression())
+                                .toString();
+                type.replaceAnnotation(
+                        qualHierarchy.greatestLowerBound(
```

Comment:
Because we're combining the new information (that we have an array's length) with any other (possibly more specific) information that we might already have about the array length. For instance, if we already know that `a.length` is LTEL("a", "b"), we don't want to overwrite that with just LTEL("a")